### ISSUE TITLE
Type of inherited attribute is lost when using dynamically constructed base class

### ISSUE THREAD

**Issue by vlashada** [astral-sh/ty#1324]
### Summary

When returning a class type (type[A]) from a function and using it as a dynamic base class, ty fails to preserve attribute type information in the subclass. The inherited attribute is recognized as unknown instead of the expected static type.

```python
from typing import reveal_type

class A:
    name: str

def make_a() -> type[A]:
    return A

class B(make_a()): # Unsupported class base with type `type[A]` [unsupported-base]
    pass

a = A()
reveal_type(a.name) # Revealed type: `str`
b = B()
reveal_type(b.name) # Revealed type: `unknown`
```

### Version

ty 0.0.1-alpha.21 (ef52a1940 2025-09-19)

**Comment by sharkdp** [comment]
Thank you for reporting this. Using `type[A]` as a type for a base class seems wrong to me? `type[A]` is a type that includes the class `A` itself, but also all subclasses of `A`. So we can't really know what a base class of type `type[A]` means. Or is this specified somewhere to be a supported typing pattern?

What you would really want to have is to return a type that *only* includes the class `A` itself. This can't be spelled out using normal Python syntax/types, though. In `ty` in particular, you could write it as `TypeOf[A]` (see [this playground](https://play.ty.dev/8ee5ef97-729e-4db2-8392-02c257afc1db)), but you probably don't want to rely on `ty_extensions` in a real code base.

**Comment by sharkdp** [comment]
For reference, both mypy and pyrefly refuse your code as well ("Unsupported dynamic base class", "Invalid expression form for base class: `make_a()`"). pyright seems to accept this code and treat it as if `A` would be the base class. However, if I recall correctly, pyright does treat `type[A]` as being synonymous with "the type that only includes the class `A` itself" in other cases as well.

**Comment by vlashada** [comment]
The type of `b.name` is still unknown when using `ty_extensions.TypeOf` as in your playground example?

**Comment by carljm** [comment]
> The type of `b.name` is still unknown when using `ty_extensions.TypeOf` as in your playground example?

No, the type of `b.name` is revealed as `str` in the playground example. Are you maybe looking at the comment in the code sample (which is out of date) instead of the actual diagnostic emitted by ty (in the lower pane)?
